archive/archive.go

```diff
@@ -0,0 +1,265 @@
+package archive
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/bitrise-io/go-utils/log"
+	"github.com/bitrise-io/go-utils/pathutil"
+	"github.com/bitrise-tools/go-xcode/plistutil"
+	"github.com/bitrise-tools/go-xcode/profileutil"
+)
+
+// Application ...
+type Application struct {
+	Path                string
+	InfoPlist           plistutil.PlistData
+	Entitlements        plistutil.PlistData
+	ProvisioningProfile profileutil.ProvisioningProfileInfoModel
+	Plugins             []Application
+}
+
+// BundleIdentifier ...
+func (app Application) BundleIdentifier() string {
+	bundleID, _ := app.InfoPlist.GetString("CFBundleIdentifier")
+	return bundleID
+}
+
+// NewApplication ...
+func NewApplication(appOrAppexPth string) (Application, error) {
+	application := Application{
+		Path: appOrAppexPth,
+	}
+
+	{
+		infoPlistPth := filepath.Join(appOrAppexPth, "Info.plist")
+		exist, err := pathutil.IsPathExists(infoPlistPth)
+		if err != nil {
+			return Application{}, err
+		} else if !exist {
+			return Application{}, fmt.Errorf("Info.plist does not exist at: %s", infoPlistPth)
+		}
+		infoPlist, err := plistutil.NewPlistDataFromFile(infoPlistPth)
+		if err != nil {
+			return Application{}, err
+		}
+		application.InfoPlist = infoPlist
+	}
+
+	{
+		provisioningProfilePth := filepath.Join(appOrAppexPth, "embedded.mobileprovision")
+		exist, err := pathutil.IsPathExists(provisioningProfilePth)
+		if err != nil {
+			return Application{}, err
+		} else if !exist {
+			return Application{}, fmt.Errorf("embedded.mobileprovision does not exist at: %s", provisioningProfilePth)
+		}
+		profile, err := profileutil.NewProvisioningProfileInfoFromFile(provisioningProfilePth)
+		if err != nil {
+			return Application{}, err
+		}
+		application.ProvisioningProfile = profile
+	}
+
+	{
+		entitlementsPth := filepath.Join(appOrAppexPth, "archived-expanded-entitlements.xcent")
+		exist, err := pathutil.IsPathExists(entitlementsPth)
+		if err != nil {
+			return Application{}, err
+		} else if exist {
+			entitlements, err := plistutil.NewPlistDataFromFile(entitlementsPth)
+			if err != nil {
+				return Application{}, err
+			}
+
+			application.Entitlements = entitlements
+		}
+	}
+
+	return application, nil
+}
+
+// Applications ...
+type Applications struct {
+	MainApplication  Application
+	WatchApplication *Application
+}
+
+// NewApplications ...
+func NewApplications(applicationsDir string) (Applications, error) {
+	mainApplication := Application{}
+	mainApplicationPth := ""
+	{
+		pattern := filepath.Join(applicationsDir, "*.app")
+		pths, err := filepath.Glob(pattern)
+		if err != nil {
+			return Applications{}, err
+		}
+
+		if len(pths) == 0 {
+			return Applications{}, fmt.Errorf("Failed to find main application using pattern: %s", pattern)
+		} else if len(pths) > 1 {
+			log.Warnf("Multiple main applications found")
```

Comment:
Instead of printing warning message in a package should we fail instantly if multiple main *.app found? Does it matter if instead of an error message we've print warning messages?